lib/appsignal/utils/argument_cleaner.ex

```diff
@@ -1,17 +1,124 @@
+defmodule Appsignal.Utils.Literal do
+  defstruct [:value]
+end
+
+defimpl Inspect, for: Appsignal.Utils.Literal do
+  def inspect(%Appsignal.Utils.Literal{value: value}, _opts), do: value
+end
+
 defmodule Appsignal.Utils.ArgumentCleaner do
-  alias Appsignal.Utils.Type
+  @maximum_iterable_count 4
+  @maximum_recursion 2
 
-  def clean(argument)
-      when is_boolean(argument) or is_integer(argument) or is_float(argument) or is_pid(argument) or
-             is_port(argument) or is_reference(argument) do
-    argument
-  end
+  def clean_literal(argument), do: %Appsignal.Utils.Literal{value: clean(argument)}
+  def clean(argument), do: clean(argument, @maximum_recursion)
 
-  def clean(argument) when is_map(argument) do
+  def clean(argument, _) when is_nil(argument), do: inspect(argument)
+  def clean(argument, _) when is_atom(argument), do: inspect(argument)
+  def clean(argument, _) when is_binary(argument), do: "\"...\""
+  def clean(argument, _) when is_bitstring(argument), do: "<<...>>"
+
+  def clean(argument, recurse) when is_map(argument) do
     {struct, map} = Map.pop(argument, :__struct__)
 
-    "%#{Appsignal.Utils.module_name(struct)}{#{Enum.map_join(map, ", ", fn {key, value} -> "#{inspect(key)} => #{inspect(clean(value))}" end)}}"
+    contents = if Enum.empty?(map), do: "", else: do_clean_map(map, recurse, struct)
+
+    if struct do
+      "%#{Appsignal.Utils.module_name(struct)}{#{contents}}"
+    else
+      "%{#{contents}}"
+    end
+  end
+
+  def clean(argument, _) when is_integer(argument), do: "integer()"
+
+  def clean(argument, _) when is_float(argument), do: "float()"
+
+  def clean(argument, _) when is_pid(argument), do: "#PID<...>"
+
+  def clean(argument, _) when is_port(argument), do: "#Port<...>"
+
+  def clean(argument, _) when is_reference(argument), do: "#Reference<...>"
+
+  def clean(argument, _) when is_function(argument) do
+    inspected_function = inspect(argument)
+
+    if String.starts_with?(inspected_function, "#Function<") do
+      # Represent the function as `fn _, _ -> ... end`, as this allows
+      # to express the arity clearly while redacting the rest.
+      arity = :erlang.fun_info(argument)[:arity]
+      placeholders = if arity == 0, do: "", else: Enum.map_join(1..arity, ", ", fn _ -> "_" end)
```

Comment:
Annoyingly, I have to check for `Enum.empty?` here because, with `1..arity`, `Enum.to_list(1..0)` results in `[1, 0]`, which is not what we want. We could use `1..arity//1` to fix that (`Enum.to_list(1..0//1) == []`) but it's not compatible with Elixir 1.11, which we still support.

I'm using `map_join/3` because it seems like the right tool for the job (and because the previous implementation used it) but I'm not a fan either -- I can never remember the order of the arguments myself.

I'll switch to the second implementation you suggest, the one that uses `map_join/3` with the separate zero-arity clause.

Reply:
Done.